README.md
[![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/kmedian/keras-layernorm-rnn/master?urlpath=lab)
[![Gitpod - Code Now](https://img.shields.io/badge/Gitpod-code%20now-blue.svg?longCache=true)](https://gitpod.io#https://github.com/kmedian/keras-layernorm-rnn)

# keras-layernorm-rnn


## Table of Contents
* [Installation](#installation)
* [Usage](#usage)
* [Commands](#commands)
* [Support](#support)
* [Contributing](#contributing)


## Installation
The `keras-layernorm-rnn` [git repo](http://github.com/kmedian/keras-layernorm-rnn) is available as [PyPi package](https://pypi.org/project/keras-layernorm-rnn)

```
pip install keras-layernorm-rnn
pip install git+ssh://git@github.com/kmedian/keras-layernorm-rnn.git
```


## Usage
Check the [examples](https://github.com/kmedian/keras-layernorm-rnn/blob/master/examples/How%20to%20use%20LayernormSimpleRNN.ipynb) folder for notebooks.


```
import tensorflow as tf
from keras_layernorm_rnn import LayernormLSTM3

model = tf.keras.Sequential([
    LayernormLSTM3(units=8, return_sequences=False),  # Many-to-One
    tf.keras.layers.Dense(1, activation='linear')
])
```

## Commands
Install a virtual environment

```
python3 -m venv .venv
source .venv/bin/activate
pip3 install --upgrade pip
pip3 install -r requirements.txt
```

(If your git repo is stored in a folder with whitespaces, then don't use the subfolder `.venv`. Use an absolute path without whitespaces.)

Python commands

* Jupyter for the examples: `jupyter lab`
* Check syntax: `flake8 --ignore=F401 --exclude=$(grep -v '^#' .gitignore | xargs | sed -e 's/ /,/g')`
* Run Unit Test: `python keras_layernorm_rnn/layernorm_simplernn_test.py`
* Upload to PyPi with twine: `python setup.py sdist && twine upload -r pypi dist/*`

Clean up 

```
find . -type f -name "*.pyc" | xargs rm
find . -type d -name "__pycache__" | xargs rm -r
rm -r .pytest_cache
rm -r .venv
```


## Support
Please [open an issue](https://github.com/kmedian/keras-layernorm-rnn/issues/new) for support.


## Contributing
Please contribute using [Github Flow](https://guides.github.com/introduction/flow/). Create a branch, add commits, and [open a pull request](https://github.com/kmedian/keras-layernorm-rnn/compare/).


CHANGES.md
# 0.2.1 / 2020-01-18

  * use_gamma flag added to disable training of gamma parameters

# 0.2.0 / 2020-01-17

  * LayernormLSTM1-3 added

# 0.1.1 / 2020-01-15

  * docstring example corrected

# 0.1.0 / 2020-01-15

  * Initial Release


keras_layernorm_rnn/layernorm_simplernn_test.py
import numpy as np
import tensorflow as tf
import tensorflow.keras as keras
from tensorflow.python.keras import testing_utils
# from tensorflow.python.framework import test_util  # deprecated

from layernorm_simplernn import LayernormSimpleRNN
# from keras_layernorm_rnn import LayernormSimpleRNN


# @test_util.run_all_in_graph_and_eager_modes
class LayernormSimpleRNNTest(tf.test.TestCase):
    def test_dynamic_behavior_layernorm_rnn(self):
        num_samples = 2
        timesteps = 3
        embedding_dim = 4
        units = 2
        layer = LayernormSimpleRNN(
            units, use_layernorm=True, input_shape=(None, embedding_dim))
        model = keras.models.Sequential()
        model.add(layer)
        model.compile('rmsprop', 'mse')
        x = np.random.random((num_samples, timesteps, embedding_dim))
        y = np.random.random((num_samples, units))
        model.train_on_batch(x, y)

    def test_constraints_layernorm_rnn(self):
        embedding_dim = 4
        layer_class = LayernormSimpleRNN
        k_constraint = keras.constraints.max_norm(0.01)
        r_constraint = keras.constraints.max_norm(0.01)
        b_constraint = keras.constraints.max_norm(0.01)
        g_constraint = keras.constraints.max_norm(0.01)
        layer = layer_class(
            5,
            use_layernorm=True,
            return_sequences=False,
            weights=None,
            input_shape=(None, embedding_dim),
            kernel_constraint=k_constraint,
            recurrent_constraint=r_constraint,
            bias_constraint=b_constraint,
            gamma_constraint=g_constraint)
        layer.build((None, None, embedding_dim))
        self.assertEqual(layer.cell.kernel.constraint, k_constraint)
        self.assertEqual(layer.cell.recurrent_kernel.constraint, r_constraint)
        self.assertEqual(layer.cell.bias.constraint, b_constraint)
        self.assertEqual(layer.cell.gamma.constraint, g_constraint)

    def test_with_masking_layer_layernorm_rnn(self):
        layer_class = LayernormSimpleRNN
        inputs = np.random.random((2, 3, 4))
        targets = np.abs(np.random.random((2, 3, 5)))
        targets /= targets.sum(axis=-1, keepdims=True)
        model = keras.models.Sequential()
        model.add(keras.layers.Masking(input_shape=(3, 4)))
        model.add(
            layer_class(
                units=5,
                use_layernorm=True,
                return_sequences=True,
                unroll=False))
        model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
        model.fit(inputs, targets, epochs=1, batch_size=2, verbose=1)

    def test_from_config_layernorm_rnn(self):
        layer_class = LayernormSimpleRNN
        for stateful in (False, True):
            l1 = layer_class(units=1, use_layernorm=True, stateful=stateful)
            l2 = layer_class.from_config(l1.get_config())
            assert l1.get_config() == l2.get_config()

    def test_regularizers_layernorm_rnn(self):
        embedding_dim = 4
        layer_class = LayernormSimpleRNN
        layer = layer_class(
            5,
            use_layernorm=True,
            return_sequences=False,
            weights=None,
            input_shape=(None, embedding_dim),
            kernel_regularizer=keras.regularizers.l1(0.01),
            recurrent_regularizer=keras.regularizers.l1(0.01),
            bias_regularizer='l2',
            gamma_regularizer='l2',
            activity_regularizer='l1')
        layer.build((None, None, 2))
        self.assertEqual(len(layer.losses), 4)

    def test_versus_simplernn(self):
        embedding_dim = 4
        timesteps = 2
        settings = {
            'units': 3,
            'bias_initializer': 'ones',
            'kernel_initializer': 'ones',
            'recurrent_initializer': 'ones'
        }
        model1 = keras.Sequential()
        model1.add(keras.layers.SimpleRNN(**settings))
        model1.build((None, None, embedding_dim))

        model2 = keras.Sequential()
        model2.add(LayernormSimpleRNN(use_layernorm=False, **settings))
        model2.build((None, None, embedding_dim))

        x = 0.5 * np.ones((1, timesteps, embedding_dim))
        y_pred1 = model1.predict(x)
        y_pred2 = model2.predict(x)
        self.assertAllEqual(y_pred1, y_pred2)


if __name__ == "__main__":
    tf.test.main()


# Excluded tests
# test_return_sequences_layernorm_rnn
# test_float64_layernorm_rnn
# test_dropout_layernorm_rnn
# test_statefulness_layernorm_rnn


keras_layernorm_rnn/layernorm_lstm1.py
import tensorflow as tf
import tensorflow.keras as keras


class LayernormLSTM1Cell(keras.layers.LSTMCell):
    """LSTM with Layer Normalization (1: more smaller multiplications)

    Notice:
        LayernormLSTM1Cell(use_layernorm=False) is the same
        as LSTMCell(implementation=1)

    References:
    [1] Hochreiter, S., Schmidhuber, J., 1997. Long short-term memory.
          Neural computation 9, 1735–1780.
    [2] Ba, Jimmy Lei, Jamie Ryan Kiros, and Geoffrey E. Hinton.
          "Layer Normalization." ArXiv:1607.06450 [Cs, Stat],
          July 21, 2016. http://arxiv.org/abs/1607.06450

    """
    def __init__(
            self,
            units,
            activation='tanh',
            recurrent_activation='hard_sigmoid',
            use_bias=True,
            use_layernorm=True,
            use_gamma=True,
            layernorm_epsilon=1e-05,
            kernel_initializer='glorot_uniform',
            recurrent_initializer='orthogonal',
            bias_initializer='zeros',
            gamma_initializer='ones',
            unit_forget_bias=True,
            kernel_regularizer=None,
            recurrent_regularizer=None,
            bias_regularizer=None,
            gamma_regularizer=None,
            kernel_constraint=None,
            recurrent_constraint=None,
            bias_constraint=None,
            gamma_constraint=None,
            dropout=0.,
            recurrent_dropout=0.,
            **kwargs):
        # store LSTM attributes
        keras.layers.LSTMCell.__init__(
            self,
            units,
            activation=activation,
            recurrent_activation=recurrent_activation,
            use_bias=use_bias,
            kernel_initializer=kernel_initializer,
            recurrent_initializer=recurrent_initializer,
            bias_initializer=bias_initializer,
            unit_forget_bias=unit_forget_bias,
            kernel_regularizer=kernel_regularizer,
            recurrent_regularizer=recurrent_regularizer,
            bias_regularizer=bias_regularizer,
            kernel_constraint=kernel_constraint,
            recurrent_constraint=recurrent_constraint,
            bias_constraint=bias_constraint,
            dropout=dropout,
            recurrent_dropout=recurrent_dropout,
            implementation=1,
            dtype=kwargs.get('dtype'),
            trainable=kwargs.get('trainable', True))
        # store layernorm attributes
        self.use_layernorm = use_layernorm
        self.use_gamma = use_gamma
        self.layernorm_epsilon = layernorm_epsilon
        self.gamma_initializer = keras.initializers.get(gamma_initializer)
        self.gamma_regularizer = keras.regularizers.get(gamma_regularizer)
        self.gamma_constraint = keras.constraints.get(gamma_constraint)

    def build(self, input_shape):
        # build input kernel, recurrent kernel and bias in LSTMCell
        keras.layers.LSTMCell.build(self, input_shape)

        # build the layernorm objects
        if self.use_layernorm:
            ln_config = {
                'axis': -1,
                'epsilon': self.layernorm_epsilon,
                'center': False,
                'scale': self.use_gamma,
                'beta_initializer': None,
                'gamma_initializer': self.gamma_initializer,
                'beta_regularizer': None,
                'gamma_regularizer': self.gamma_regularizer,
                'beta_constraint': None,
                'gamma_constraint': self.gamma_constraint,
                'dtype': self.dtype,
                'trainable': self.trainable
            }
            self.layernorm_i = keras.layers.LayerNormalization(**ln_config)
            self.layernorm_f = keras.layers.LayerNormalization(**ln_config)
            self.layernorm_c = keras.layers.LayerNormalization(**ln_config)
            self.layernorm_o = keras.layers.LayerNormalization(**ln_config)

    def call(self, inputs, states, training=None):
        # read results from previous time step
        h_tm1 = states[0]  # previous memory state
        c_tm1 = states[1]  # previous carry state

        # IMPLEMENTATION 1
        # - Fewer small operations instead of few big operations

        # dropout, input kernel
        if (training is not None) and (0. < self.dropout < 1.):
            # generate 4x dropout masks from DropoutRNNCellMixin
            dp_mask = self.get_dropout_mask_for_cell(
                inputs, training, count=4)
            # apply dropouts
            inputs_i = inputs * dp_mask[0]
            inputs_f = inputs * dp_mask[1]
            inputs_c = inputs * dp_mask[2]
            inputs_o = inputs * dp_mask[3]
        else:
            inputs_i, inputs_f, inputs_c, inputs_o = (
                inputs, inputs, inputs, inputs)

        # multiply inputs with weight matrix
        W_i, W_f, W_c, W_o = tf.split(
            self.kernel, num_or_size_splits=4, axis=1)
        net_i = keras.backend.dot(inputs_i, W_i)
        net_f = keras.backend.dot(inputs_f, W_f)
        net_c = keras.backend.dot(inputs_c, W_c)
        net_o = keras.backend.dot(inputs_o, W_o)

        # dropout, recurrent kernel
        if (training is not None) and (0. < self.recurrent_dropout < 1.):
            # generate 4x dropout masks from DropoutRNNCellMixin
            rec_dp_mask = self.get_recurrent_dropout_mask_for_cell(
                h_tm1, training, count=4)
            # apply dropouts
            h_tm1_i = h_tm1 * rec_dp_mask[0]
            h_tm1_f = h_tm1 * rec_dp_mask[1]
            h_tm1_c = h_tm1 * rec_dp_mask[2]
            h_tm1_o = h_tm1 * rec_dp_mask[3]
        else:
            h_tm1_i, h_tm1_f, h_tm1_c, h_tm1_o = h_tm1, h_tm1, h_tm1, h_tm1

        # multiply previous hidden with recurrent weights
        R_i, R_f, R_c, R_o = tf.split(
            self.recurrent_kernel, num_or_size_splits=4, axis=1)
        net_i += keras.backend.dot(h_tm1_i, R_i)
        net_f += keras.backend.dot(h_tm1_f, R_f)
        net_c += keras.backend.dot(h_tm1_c, R_c)
        net_o += keras.backend.dot(h_tm1_o, R_o)

        # apply scaling of layer normalization to each gate
        if self.use_layernorm:
            net_i = self.layernorm_i(net_i)
            net_f = self.layernorm_f(net_f)
            net_c = self.layernorm_c(net_c)
            net_o = self.layernorm_o(net_o)

        # apply bias
        if self.use_bias:
            bias_i, bias_f, bias_c, bias_o = tf.split(
                self.bias, num_or_size_splits=4, axis=0)
            net_i = keras.backend.bias_add(net_i, bias_i)
            net_f = keras.backend.bias_add(net_f, bias_f)
            net_c = keras.backend.bias_add(net_c, bias_c)
            net_o = keras.backend.bias_add(net_o, bias_o)

        # _compute_carry_and_output_fused
        i = self.recurrent_activation(net_i)
        f = self.recurrent_activation(net_f)
        c = f * c_tm1 + i * self.activation(net_c)
        o = self.recurrent_activation(net_o)

        # merge all signals into memory state h
        h = o * self.activation(c)

        # return memory state h, and carry state c
        return h, [h, c]

    def get_config(self):
        config = {
            'use_layernorm': self.use_layernorm,
            'use_gamma': self.use_gamma,
            'layernorm_epsilon': self.layernorm_epsilon,
            'gamma_initializer': self.gamma_initializer,
            'gamma_regularizer': self.gamma_regularizer,
            'gamma_constraint': self.gamma_constraint
        }

        cell_config = keras.layers.LSTMCell.get_config(self)
        del cell_config['name']

        return {**config, **cell_config}


class LayernormLSTM1(keras.layers.LSTM):
    def __init__(
            self,
            units,
            activation='tanh',
            recurrent_activation='hard_sigmoid',
            use_bias=True,
            use_layernorm=True,
            use_gamma=True,
            layernorm_epsilon=1e-05,
            kernel_initializer='glorot_uniform',
            recurrent_initializer='orthogonal',
            bias_initializer='zeros',
            gamma_initializer='ones',
            unit_forget_bias=True,
            kernel_regularizer=None,
            recurrent_regularizer=None,
            bias_regularizer=None,
            gamma_regularizer=None,
            activity_regularizer=None,
            kernel_constraint=None,
            recurrent_constraint=None,
            bias_constraint=None,
            gamma_constraint=None,
            dropout=0.,
            recurrent_dropout=0.,
            return_sequences=False,
            return_state=False,
            go_backwards=False,
            stateful=False,
            unroll=False,
            **kwargs):
        # instantiate the RNN cell
        cell = LayernormLSTM1Cell(
            units,
            activation=activation,
            recurrent_activation=recurrent_activation,
            use_bias=use_bias,
            use_layernorm=use_layernorm,
            use_gamma=use_gamma,
            layernorm_epsilon=layernorm_epsilon,
            kernel_initializer=kernel_initializer,
            recurrent_initializer=recurrent_initializer,
            bias_initializer=bias_initializer,
            gamma_initializer=gamma_initializer,
            unit_forget_bias=unit_forget_bias,
            kernel_regularizer=kernel_regularizer,
            recurrent_regularizer=recurrent_regularizer,
            bias_regularizer=bias_regularizer,
            gamma_regularizer=gamma_regularizer,
            activity_regularizer=activity_regularizer,
            kernel_constraint=kernel_constraint,
            recurrent_constraint=recurrent_constraint,
            bias_constraint=bias_constraint,
            gamma_constraint=gamma_constraint,
            dropout=dropout,
            recurrent_dropout=recurrent_dropout,
            dtype=kwargs.get('dtype'),
            trainable=kwargs.get('trainable', True))
        # call the parent class 'RNN'
        keras.layers.RNN.__init__(
            self,
            cell,
            return_sequences=return_sequences,
            return_state=return_state,
            go_backwards=go_backwards,
            stateful=stateful,
            unroll=unroll,
            **kwargs)
        # set other parameters
        self.activity_regularizer = keras.regularizers.get(
            activity_regularizer)
        self.input_spec = [keras.layers.InputSpec(ndim=3)]
        # no cuDNN supoort
        self.could_use_cudnn = False
        # return_runtime is a flag for testing (grappler, graph mode)
        self.return_runtime = kwargs.pop('return_runtime', False)

    @property
    def use_layernorm(self):
        return self.cell.use_layernorm

    @property
    def use_gamma(self):
        return self.cell.use_gamma

    @property
    def layernorm_epsilon(self):
        return self.cell.layernorm_epsilon

    @property
    def gamma_initializer(self):
        return self.cell.gamma_initializer

    @property
    def gamma_regularizer(self):
        return self.cell.gamma_regularizer

    @property
    def gamma_constraint(self):
        return self.cell.gamma_constraint

    def get_config(self):
        base_config = keras.layers.RNN.get_config(self)
        del base_config['cell']

        cell_config = self.cell.get_config()
        return {**base_config, **cell_config}


keras_layernorm_rnn/layernorm_lstm3.py
import tensorflow as tf
import tensorflow.keras as keras


class LayernormLSTM3Cell(keras.layers.LSTMCell):
    """LSTM with Layer Normalization

    Notice:
        LayernormLSTM3Cell(use_layernorm=False) is the same
        as LSTMCell(implementation=1)

    References:
    [1] Hochreiter, S., Schmidhuber, J., 1997. Long short-term memory.
          Neural computation 9, 1735–1780.
    [2] Ba, Jimmy Lei, Jamie Ryan Kiros, and Geoffrey E. Hinton.
          "Layer Normalization." ArXiv:1607.06450 [Cs, Stat],
          July 21, 2016. http://arxiv.org/abs/1607.06450

    """
    def __init__(
            self,
            units,
            activation='tanh',
            recurrent_activation='hard_sigmoid',
            use_bias=True,
            use_layernorm=True,
            use_gamma=True,
            layernorm_epsilon=1e-05,
            kernel_initializer='glorot_uniform',
            recurrent_initializer='orthogonal',
            bias_initializer='zeros',
            gamma_initializer='ones',
            unit_forget_bias=True,
            kernel_regularizer=None,
            recurrent_regularizer=None,
            bias_regularizer=None,
            gamma_regularizer=None,
            kernel_constraint=None,
            recurrent_constraint=None,
            bias_constraint=None,
            gamma_constraint=None,
            dropout=0.,
            recurrent_dropout=0.,
            **kwargs):
        # store LSTM attributes
        keras.layers.LSTMCell.__init__(
            self,
            units,
            activation=activation,
            recurrent_activation=recurrent_activation,
            use_bias=use_bias,
            kernel_initializer=kernel_initializer,
            recurrent_initializer=recurrent_initializer,
            bias_initializer=bias_initializer,
            unit_forget_bias=unit_forget_bias,
            kernel_regularizer=kernel_regularizer,
            recurrent_regularizer=recurrent_regularizer,
            bias_regularizer=bias_regularizer,
            kernel_constraint=kernel_constraint,
            recurrent_constraint=recurrent_constraint,
            bias_constraint=bias_constraint,
            dropout=dropout,
            recurrent_dropout=recurrent_dropout,
            implementation=1,
            dtype=kwargs.get('dtype'),
            trainable=kwargs.get('trainable', True))
        # store layernorm attributes
        self.use_layernorm = use_layernorm
        self.use_gamma = use_gamma
        self.layernorm_epsilon = layernorm_epsilon
        self.gamma_initializer = keras.initializers.get(gamma_initializer)
        self.gamma_regularizer = keras.regularizers.get(gamma_regularizer)
        self.gamma_constraint = keras.constraints.get(gamma_constraint)

    def build(self, input_shape):
        # get sizes
        input_dim = input_shape[-1]

        # Input weight matrices W_i, W_f, W_c, W_o
        w_config = {
            'shape': (input_dim, self.units),
            'initializer': self.kernel_initializer,
            'regularizer': self.kernel_regularizer,
            'constraint': self.kernel_constraint
        }
        self.W_i = self.add_weight(**w_config, name="W_i")
        self.W_f = self.add_weight(**w_config, name="W_f")
        self.W_c = self.add_weight(**w_config, name="W_c")
        self.W_o = self.add_weight(**w_config, name="W_o")

        # Recurrent weight matrices R_i, R_f, R_c, R_o
        r_config = {
            'shape': (self.units, self.units),
            'initializer': self.recurrent_initializer,
            'regularizer': self.recurrent_regularizer,
            'constraint': self.recurrent_constraint
        }
        self.R_i = self.add_weight(**r_config, name="R_i")
        self.R_f = self.add_weight(**r_config, name="R_f")
        self.R_c = self.add_weight(**r_config, name="R_c")
        self.R_o = self.add_weight(**r_config, name="R_o")

        # Bias
        if self.use_bias:
            # Change defaults
            if self.unit_forget_bias:
                b_config = {
                    'shape': (self.units, ),
                    'regularizer': self.bias_regularizer,
                    'constraint': self.bias_constraint
                }
                self.b_i = self.add_weight(
                    **b_config, initializer=self.bias_initializer, name="b_i")
                self.b_f = self.add_weight(
                    **b_config, initializer='ones', name="b_f")
                self.b_c = self.add_weight(
                    **b_config, initializer=self.bias_initializer, name="b_c")
                self.b_o = self.add_weight(
                    **b_config, initializer=self.bias_initializer, name="b_o")
            else:
                b_config = {
                    'shape': (self.units, ),
                    'initializer': self.bias_initializer,
                    'regularizer': self.bias_regularizer,
                    'constraint': self.bias_constraint
                }
                self.b_i = self.add_weight(**b_config, name="b_i")
                self.b_f = self.add_weight(**b_config, name="b_f")
                self.b_c = self.add_weight(**b_config, name="b_c")
                self.b_o = self.add_weight(**b_config, name="b_o")

        # build the layernorm objects
        if self.use_layernorm:
            ln_config = {
                'axis': -1,
                'epsilon': self.layernorm_epsilon,
                'center': False,
                'scale': self.use_gamma,
                'beta_initializer': None,
                'gamma_initializer': self.gamma_initializer,
                'beta_regularizer': None,
                'gamma_regularizer': self.gamma_regularizer,
                'beta_constraint': None,
                'gamma_constraint': self.gamma_constraint,
                'dtype': self.dtype,
                'trainable': self.trainable
            }
            self.layernorm_i = keras.layers.LayerNormalization(**ln_config)
            self.layernorm_f = keras.layers.LayerNormalization(**ln_config)
            self.layernorm_c = keras.layers.LayerNormalization(**ln_config)
            self.layernorm_o = keras.layers.LayerNormalization(**ln_config)

    def call(self, inputs, states, training=None):
        # read results from previous time step
        h_tm1 = states[0]  # previous memory state
        c_tm1 = states[1]  # previous carry state

        # IMPLEMENTATION 1
        # - Fewer small operations instead of few big operations

        # dropout, input kernel
        if (training is not None) and (0. < self.dropout < 1.):
            # generate 4x dropout masks from DropoutRNNCellMixin
            dp_mask = self.get_dropout_mask_for_cell(
                inputs, training, count=4)
            # apply dropouts
            inputs_i = inputs * dp_mask[0]
            inputs_f = inputs * dp_mask[1]
            inputs_c = inputs * dp_mask[2]
            inputs_o = inputs * dp_mask[3]
        else:
            inputs_i, inputs_f, inputs_c, inputs_o = (
                inputs, inputs, inputs, inputs)

        # multiply inputs with weight matrix
        net_i = keras.backend.dot(inputs_i, self.W_i)
        net_f = keras.backend.dot(inputs_f, self.W_f)
        net_c = keras.backend.dot(inputs_c, self.W_c)
        net_o = keras.backend.dot(inputs_o, self.W_o)

        # dropout, recurrent kernel
        if (training is not None) and (0. < self.recurrent_dropout < 1.):
            # generate 4x dropout masks from DropoutRNNCellMixin
            rec_dp_mask = self.get_recurrent_dropout_mask_for_cell(
                h_tm1, training, count=4)
            # apply dropouts
            h_tm1_i = h_tm1 * rec_dp_mask[0]
            h_tm1_f = h_tm1 * rec_dp_mask[1]
            h_tm1_c = h_tm1 * rec_dp_mask[2]
            h_tm1_o = h_tm1 * rec_dp_mask[3]
        else:
            h_tm1_i, h_tm1_f, h_tm1_c, h_tm1_o = h_tm1, h_tm1, h_tm1, h_tm1

        # multiply previous hidden with recurrent weights
        net_i += keras.backend.dot(h_tm1_i, self.R_i)
        net_f += keras.backend.dot(h_tm1_f, self.R_f)
        net_c += keras.backend.dot(h_tm1_c, self.R_c)
        net_o += keras.backend.dot(h_tm1_o, self.R_o)

        # apply scaling of layer normalization to each gate
        if self.use_layernorm:
            net_i = self.layernorm_i(net_i)
            net_f = self.layernorm_f(net_f)
            net_c = self.layernorm_c(net_c)
            net_o = self.layernorm_o(net_o)

        # apply bias
        if self.use_bias:
            net_i = keras.backend.bias_add(net_i, self.b_i)
            net_f = keras.backend.bias_add(net_f, self.b_f)
            net_c = keras.backend.bias_add(net_c, self.b_c)
            net_o = keras.backend.bias_add(net_o, self.b_o)

        # _compute_carry_and_output_fused
        i = self.recurrent_activation(net_i)
        f = self.recurrent_activation(net_f)
        c = f * c_tm1 + i * self.activation(net_c)
        o = self.recurrent_activation(net_o)

        # merge all signals into memory state h
        h = o * self.activation(c)

        # return memory state h, and carry state c
        return h, [h, c]

    def get_config(self):
        config = {
            'use_layernorm': self.use_layernorm,
            'use_gamma': self.use_gamma,
            'layernorm_epsilon': self.layernorm_epsilon,
            'gamma_initializer': self.gamma_initializer,
            'gamma_regularizer': self.gamma_regularizer,
            'gamma_constraint': self.gamma_constraint
        }

        cell_config = keras.layers.LSTMCell.get_config(self)
        del cell_config['name']

        return {**config, **cell_config}


class LayernormLSTM3(keras.layers.LSTM):
    def __init__(
            self,
            units,
            activation='tanh',
            recurrent_activation='hard_sigmoid',
            use_bias=True,
            use_layernorm=True,
            use_gamma=True,
            layernorm_epsilon=1e-05,
            kernel_initializer='glorot_uniform',
            recurrent_initializer='orthogonal',
            bias_initializer='zeros',
            gamma_initializer='ones',
            unit_forget_bias=True,
            kernel_regularizer=None,
            recurrent_regularizer=None,
            bias_regularizer=None,
            gamma_regularizer=None,
            activity_regularizer=None,
            kernel_constraint=None,
            recurrent_constraint=None,
            bias_constraint=None,
            gamma_constraint=None,
            dropout=0.,
            recurrent_dropout=0.,
            return_sequences=False,
            return_state=False,
            go_backwards=False,
            stateful=False,
            unroll=False,
            **kwargs):
        # instantiate the RNN cell
        cell = LayernormLSTM3Cell(
            units,
            activation=activation,
            recurrent_activation=recurrent_activation,
            use_bias=use_bias,
            use_layernorm=use_layernorm,
            use_gamma=use_gamma,
            layernorm_epsilon=layernorm_epsilon,
            kernel_initializer=kernel_initializer,
            recurrent_initializer=recurrent_initializer,
            bias_initializer=bias_initializer,
            gamma_initializer=gamma_initializer,
            unit_forget_bias=unit_forget_bias,
            kernel_regularizer=kernel_regularizer,
            recurrent_regularizer=recurrent_regularizer,
            bias_regularizer=bias_regularizer,
            gamma_regularizer=gamma_regularizer,
            activity_regularizer=activity_regularizer,
            kernel_constraint=kernel_constraint,
            recurrent_constraint=recurrent_constraint,
            bias_constraint=bias_constraint,
            gamma_constraint=gamma_constraint,
            dropout=dropout,
            recurrent_dropout=recurrent_dropout,
            dtype=kwargs.get('dtype'),
            trainable=kwargs.get('trainable', True))
        # call the parent class 'RNN'
        keras.layers.RNN.__init__(
            self,
            cell,
            return_sequences=return_sequences,
            return_state=return_state,
            go_backwards=go_backwards,
            stateful=stateful,
            unroll=unroll,
            **kwargs)
        # set other parameters
        self.activity_regularizer = keras.regularizers.get(
            activity_regularizer)
        self.input_spec = [keras.layers.InputSpec(ndim=3)]
        # no cuDNN supoort
        self.could_use_cudnn = False
        # return_runtime is a flag for testing (grappler, graph mode)
        self.return_runtime = kwargs.pop('return_runtime', False)

    @property
    def use_layernorm(self):
        return self.cell.use_layernorm

    @property
    def use_gamma(self):
        return self.cell.use_gamma

    @property
    def layernorm_epsilon(self):
        return self.cell.layernorm_epsilon

    @property
    def gamma_initializer(self):
        return self.cell.gamma_initializer

    @property
    def gamma_regularizer(self):
        return self.cell.gamma_regularizer

    @property
    def gamma_constraint(self):
        return self.cell.gamma_constraint

    def get_config(self):
        base_config = keras.layers.RNN.get_config(self)
        del base_config['cell']

        cell_config = self.cell.get_config()
        return {**base_config, **cell_config}
